File: ts/src/data/gamedataAttrs.ts
/**
 * Shared GameData attribute helpers.
 *
 * Holds the mValue unwrap used by enemy stats and stage-enemy level parsing;
 * previously duplicated across enemy.ts and stageEnemy.ts. Mirrors
 * python/src/prts_mcp/data/gamedata_attrs.py.
 */

export interface MValue {
  m_defined?: boolean;
  m_value?: unknown;
}

/** Unwrap {m_defined, m_value} if present, else return as-is. */
export function mValue<T>(obj: unknown, defaultValue: T): T;
export function mValue<T>(obj: unknown, defaultValue?: T): T | undefined;
export function mValue<T>(obj: unknown, defaultValue?: T): T | unknown {
  if (obj && typeof obj === "object" && "m_value" in obj) {
    return (obj as MValue).m_value;
  }
  return obj ?? defaultValue;
}

File: python/src/prts_mcp/data/gamedata_attrs.py
"""Shared GameData attribute helpers.

Holds the ``m_value`` unwrap used by enemy stats and stage-enemy level
parsing; previously duplicated across ``data/enemy`` and ``data/stage_enemy``.
The TS mirror is ``ts/src/data/gamedataAttrs.ts``.
"""
from __future__ import annotations

from typing import Any


def m_value(obj: Any, default: Any = None) -> Any:
    """Unwrap {m_defined, m_value} if present, else return as-is."""
    if isinstance(obj, dict) and "m_value" in obj:
        return obj["m_value"]
    return obj if obj is not None else default
